---
title: "CT3 & others, compare annotation rate"
output: html_notebook
---

load packages
```{r}
library(ggplot2)
library(data.table)
library(dplyr)
library(tidyr)
library(stringr)
library(wesanderson)
library(ggpubr)
library(rprojroot)
```

set file paths
```{r}
phar_dir <- sprintf(
  "%s/data/annotation/gene_to_contig/pharokka",
  find_rstudio_root_file())

ct3_dir <- sprintf(
  "%s/data/annotation/gene_to_contig/ct3",
  find_rstudio_root_file())

genomad_dir <- sprintf(
  "%s/data/annotation/gene_to_contig/genomad",
  find_rstudio_root_file())
```

load and combine tables by tool
pharokka
```{r}
if (exists("phar_dataset")){
  rm(phar_dataset)
}

#file_list <- Sys.glob("*report.txt")

## list of relevant files in the directory
phar_files <- list.files(phar_dir, pattern = "*.tsv", full.names = TRUE)



for (file in phar_files){
      
  # if the merged dataset doesn't exist, create it
  if (!exists("phar_dataset")){
    
    tax_label <- basename(file)
    tax_label <- str_split_i(tax_label, "\\.", 1)[[1]]
    
    phar_dataset <- fread(file, header=T, sep="\t") %>%
      group_by(contig) %>%
      summarize(gene_count = n(),
                hypo_annotations = 
                  n_distinct(gene[annot == "hypothetical protein"])
      ) %>%
      mutate(tool = "pharokka",
             settings = "-m -s -g prodigal-gv --meta_hmm",
             tax_label = tax_label)

  }
  
  # if the merged dataset does exist, append to it
  else if (exists("phar_dataset")){
    
    tax_label <- basename(file)
    tax_label <- str_split_i(tax_label, "\\.", 1)[[1]]
    
    temp_dataset <- fread(file, header=T, sep="\t") %>%
      group_by(contig) %>%
      summarize(gene_count = n(),
                hypo_annotations = 
                  n_distinct(gene[annot == "hypothetical protein"])
      ) %>%
      mutate(tool = "pharokka",
             settings = "-m -s -g prodigal-gv --meta_hmm",
             tax_label = tax_label)

    phar_dataset<-rbind(phar_dataset, temp_dataset)
    rm(temp_dataset)
  }
}

```


ct3
```{r}
if (exists("ct3_dataset")){
  rm(ct3_dataset)
}

#file_list <- Sys.glob("*report.txt")

## list of relevant files in the directory
ct3_files <- list.files(ct3_dir, pattern = "*.tsv", full.names = TRUE)



for (file in ct3_files){
      
  # if the merged dataset doesn't exist, create it
  if (!exists("ct3_dataset")){
    
    tax_label <- basename(file)
    tax_label <- str_split_i(tax_label, "\\.", 1)[[1]]
    
    ct3_dataset <- fread(file, header=T, sep="\t") %>%
      filter(Evidence_source != "tRNAscan-SE") %>%
      group_by(contig) %>%
      summarize(gene_count = n(),
                hypo_annotations = 
                  n_distinct(gene_name[
                    evidence_description == "hypothetical protein"
                    ]
                    )) %>%
      mutate(tool = "Cenote-Taker 3",
             settings = "-am T --caller prodigal-gv",
                 tax_label = tax_label)

  }
  
  # if the merged dataset does exist, append to it
  else if (exists("ct3_dataset")){
    
    tax_label <- basename(file)
    tax_label <- str_split_i(tax_label, "\\.", 1)[[1]]
    
    temp_dataset <- fread(file, header=T, sep="\t") %>%
      filter(Evidence_source != "tRNAscan-SE") %>%
      group_by(contig) %>%
      summarize(gene_count = n(),
                hypo_annotations = 
                  n_distinct(gene_name[
                    evidence_description == "hypothetical protein"
                    ]
                    )) %>%
      mutate(tool = "Cenote-Taker 3",
             settings = "-am T --caller prodigal-gv",
                 tax_label = tax_label)

    ct3_dataset<-rbind(ct3_dataset, temp_dataset)
    rm(temp_dataset)
  }
}

```



combine and compare
```{r}
comb_dt <- rbind(ct3_dataset, phar_dataset)
```


```{r}

pal <- wes_palette("FantasticFox1", 2, type = "discrete")
annotp <- comb_dt %>%
  ggplot(aes(x = tool, y = 1 - (hypo_annotations/gene_count),
             fill = tool)) +
  geom_boxplot(outlier.shape = NA) + 
  geom_point(color = "grey30") +
  scale_fill_manual(values = pal) + 
  theme_bw() +
  scale_y_continuous(labels = scales::percent) +
  labs(y = "Percentage of genes with\nnon-hypothetical annotation",
       x = "") +
  stat_compare_means(label = "p.signif") +
  facet_wrap(~tax_label, nrow = 1) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))

annotp

#ggsave(prod100_annotp,
#       file = "/Users/michaeltisza/mike_tisza/sandbox/annotation_analyses1/annot#_rate_2tools_prod100.pdf")
```

```{r}
caudo100_comb_dt %>%
  group_by(tool) %>%
  summarize(annot_avg = 
              scales::percent(
                mean(
                  1 - (hypo_annotations/gene_count)),
                accuracy = 0.1)
            )
```


combine and compare
```{r}
phar_meffect_comb_dt <- rbind(sum_phar_caudo_dt, sum_pharmonly_caudo_dt)
```


```{r}
pharm_annotp <- phar_meffect_comb_dt %>%
  ggplot(aes(x = settings, y = 1 - (hypo_annotations/gene_count),
             fill = settings)) +
  geom_boxplot() + 
  geom_point(color = "grey30") +
  #scale_fill_manual(values = pal) + 
  theme_bw() +
  scale_y_continuous(labels = scales::percent) +
  labs(y = "Percentage of genes with\nnon-hypothetical annotation",
       x = "Pharokka settings") +
  stat_compare_means(paired = T)

pharm_annotp

#ggsave(prod100_annotp,
#       file = "/Users/michaeltisza/mike_tisza/sandbox/annotation_analyses1/annot#_rate_2tools_prod100.pdf")
```

```{r}
phar_meffect_comb_dt %>%
  group_by(settings) %>%
  summarize(annot_avg = scales::percent(mean(1 - (hypo_annotations/gene_count)),
                accuracy = 0.1))
```









