---
title: "Seawater DTRs compare structural annoations"
output: html_notebook
---

```{r}
library(data.table)
library(stringr)
library(dplyr)
library(tidyr)
library(ggplot2)
library(forcats)
library(rprojroot)
library(ggpubr)
library(cowplot)
```



```{r}
metc_gene_dt <- fread(sprintf(
  "%s/data/annotation/seawater_dtrs/metc/final_annotation_summary.tsv",
  find_rstudio_root_file()
  )
)

gnmd_gene_dt <- fread(sprintf(
  "%s/data/annotation/seawater_dtrs/genomad/seawater_dtr_headtail_contigs_genes.tsv",
  find_rstudio_root_file()
  )
)

phar_gene_dt <- fread(sprintf(
 "%s/data/annotation/seawater_dtrs/pharokka/pharokka_cds_final_merged_output.tsv",
  find_rstudio_root_file()
  )
)
ct3_gene_dt<- fread(sprintf(
  "%s/data/annotation/seawater_dtrs/ct3/final_genes_to_contigs_annotation_summary.tsv",
  find_rstudio_root_file()
  )
)

ct3_sum_dt<- fread(sprintf(
  "%s/data/annotation/seawater_dtrs/ct3/ct3_s_dtr_ht_1a_virus_summary.tsv",
  find_rstudio_root_file()
  )
)

phold_gene_dt <- fread(sprintf(
  "%s/data/annotation/seawater_dtrs/phold/phold_per_cds_predictions.tsv",
  find_rstudio_root_file()
  )
)

```

```{r}
ct3_merge_dt <- merge(ct3_gene_dt, ct3_sum_dt %>% 
                        select(c(contig, input_name, 
                                 virion_hallmark_genes,
                                 virion_hallmark_count)),
                      by = "contig")


ct3_struct_dt <- ct3_merge_dt %>%
  rename(contigID = input_name) %>%
  filter(grepl("capsid|terminase|portal|major head protein",
               evidence_description,
               ignore.case = T),
         !grepl("minor|small",
               evidence_description,
               ignore.case = T)
) %>%
  select(c(contigID, evidence_description)) %>%
  mutate(software = "Cenote-Taker 3") %>%
  group_by(contigID, software) %>%
  summarize(n_capsid = sum(grepl("capsid|major head protein", 
                                   evidence_description,
                                   ignore.case = T)),
            n_terminase = sum(grepl("terminase", 
                                   evidence_description,
                                   ignore.case = T)),
            n_portal = sum(grepl("portal", 
                                   evidence_description,
                                   ignore.case = T)))

```




```{r}
ct3_gene_freq <- ct3_merge_dt %>%
  group_by(evidence_description) %>%
  summarize(n = n()) %>%
  arrange(desc(n))
```


```{r}

gnmd_struct_dt <- gnmd_gene_dt %>%
  mutate(contigID = sub("_[^_]+$", "", gene)) %>%
  filter(grepl("capsid|terminase|portal|major head protein|Bacteriophage head to tail connecting protein",
               annotation_description,
               ignore.case = T),
         !grepl("minor|small",
               annotation_description,
               ignore.case = T)
) %>%
  select(c(contigID, annotation_description)) %>%
  mutate(software = "geNomad") %>%
  group_by(contigID, software) %>%
  summarize(n_capsid = sum(grepl("capsid|major head protein", 
                                   annotation_description,
                                   ignore.case = T)),
            n_terminase = sum(grepl("terminase", 
                                   annotation_description,
                                   ignore.case = T)),
            n_portal = sum(grepl("portal|head to tail connecting protein", 
                                   annotation_description,
                                   ignore.case = T)))

```


```{r}
phar_struct_dt <-phar_gene_dt %>%
  rename(contigID = contig) %>%
  filter(grepl("capsid|terminase|portal|major head protein",
               annot,
               ignore.case = T),
         !grepl("minor|small",
               annot,
               ignore.case = T)
) %>%
  select(c(contigID, annot)) %>%
  mutate(software = "Pharokka") %>%
  group_by(contigID, software) %>%
  summarize(n_capsid = sum(grepl("capsid|major head protein", 
                                   annot,
                                   ignore.case = T)),
            n_terminase = sum(grepl("terminase", 
                                   annot,
                                   ignore.case = T)),
            n_portal = sum(grepl("portal", 
                                   annot,
                                   ignore.case = T)))

```


```{r}
metc_info_dt <- metc_gene_dt %>%
  mutate(contigID = sub("_[^_]+$", "", target))

metc_struct_dt <- metc_info_dt %>%
  filter(grepl("capsid|terminase|portal|major head protein|head to tail connecting protein",
               product,
               ignore.case = T),
         !grepl("minor|small",
               product,
               ignore.case = T)
) %>%
  select(c(contigID, product)) %>%
  mutate(software = "Metacerberus") %>%
  group_by(contigID, software) %>%
  summarize(n_capsid = sum(grepl("capsid|major head protein", 
                                   product,
                                   ignore.case = T)),
            n_terminase = sum(grepl("terminase", 
                                   product,
                                   ignore.case = T)),
            n_portal = sum(grepl("portal|head to tail connecting protein", 
                                   product,
                                   ignore.case = T)))

```

phold
```{r}
phold_struct_dt <- phold_gene_dt %>%
  mutate(contigID = contig_id) %>%
  filter(grepl("capsid|terminase|portal|major head protein",
               product,
               ignore.case = T),
         !grepl("minor|small",
               product,
               ignore.case = T)
) %>%
  select(c(contigID, product)) %>%
  mutate(software = "phold") %>%
  group_by(contigID, software) %>%
  summarize(n_capsid = sum(grepl("capsid|major head protein", 
                                   product,
                                   ignore.case = T)),
            n_terminase = sum(grepl("terminase", 
                                   product,
                                   ignore.case = T)),
            n_portal = sum(grepl("portal", 
                                   product,
                                   ignore.case = T)))
```

combine and complete
```{r}
sum_structl <- list(ct3_struct_dt, gnmd_struct_dt, 
                    metc_struct_dt, phar_struct_dt,
                    phold_struct_dt)

struct_all_dt <- rbindlist(sum_structl) %>%
  complete(contigID, software,
           fill = list(n_capsid = 0,
                       n_terminase = 0,
                       n_portal = 0)) %>%
  rename(MCP = n_capsid,
         TerL = n_terminase,
         Portal = n_portal) %>%
  pivot_longer(cols = c("MCP", "TerL", "Portal"),
               names_to = "structure",
               values_to = "number_annotated")
```




```{r}
every_nth = function(n) {
  return(function(x) {x[c(TRUE, rep(FALSE, n - 1))]})
}

struct_p <- struct_all_dt %>%
  mutate(contig_ab = gsub(".*_NODE", "NODE", contigID),
         contig_ab = gsub("_cov_.*", "", contig_ab),
         software = fct_relevel(software,  "geNomad",
                                           "Metacerberus", 
                                           "Pharokka", 
                                           "Cenote-Taker 3", 
                                           "phold"),
         number_annotated = as.factor(number_annotated)) %>%
  ggplot(aes(x = structure, y = contig_ab, fill = number_annotated)) +
  geom_tile() +
  facet_wrap(~software, 
             nrow = 1) +
  theme_bw() +
  scale_fill_manual(
    values = c("0" = 'grey95', "1" = "#91D5DE", "2" = "#2E8289", "3" = "#103326", "4" = "black"), 
    name = "number\nannotated"
    ) +
  labs(x = "",
       y = "Genomes",
       #title = "Seawater Virome Complete (Circular) Head-Tail Viruses"
       ) +
  theme(
    axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1),
    axis.text.y = element_blank(),
    legend.position = "bottom"
    ) 

struct_p

ggsave(
  struct_p,
  file = sprintf("%s/charts/seawater_structure_annotation_tile3.pdf",
                 find_rstudio_root_file()),
  width = 8,
  height = 6.5
)

ggsave(
  struct_p,
  file = sprintf("%s/charts/seawater_structure_annotation_tile3.png",
                 find_rstudio_root_file()),
  width = 7.5,
  height = 6.5
)
```

count "perfect" MCP/TerL/Portal annotations 
```{r}
essential_perf_dt <- rbindlist(sum_structl) %>%
  complete(contigID, software,
           fill = list(n_capsid = 0,
                       n_terminase = 0,
                       n_portal = 0)) %>%
  rename(MCP = n_capsid,
         TerL = n_terminase,
         Portal = n_portal) %>%
  mutate(perfect = case_when(
    MCP ==1 & TerL == 1 & Portal ==1 ~ TRUE,
    TRUE ~ FALSE
  )
  ) %>%
  filter(perfect == TRUE) %>%
  group_by(software) %>%
  summarize(n_perfect = n())

essential_p <- essential_perf_dt %>%
  mutate(software = fct_relevel(software,  "geNomad",
                                           "Metacerberus", 
                                           "Pharokka", 
                                           "Cenote-Taker 3", 
                                           "phold")) %>%
  ggplot(aes(y = n_perfect, x = software, fill= software)) +
  geom_col() +
  geom_text(aes(y = n_perfect, label = n_perfect),
            nudge_y = 5) +
  scale_fill_manual(values = 
                      c("#0A9F9D", "#CEB175", 
                        "#E54E21", "#6C8645", 
                        "#591f6e")) + 
  theme_bw() +
  labs(
    y = "Single Copy MCP/TerL/Portal",
    x = ""
  ) +
  theme(
    legend.position = "none",
    axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)
    )

essential_p

ggsave(
  essential_p,
  file = sprintf(
    "%s/charts/seawater_essential_genes_perfect.pdf",
    find_rstudio_root_file()
  ),
  height = 5,
  width = 3
)

ggsave(
  essential_p,
  file = sprintf(
    "%s/charts/seawater_essential_genes_perfect.png",
    find_rstudio_root_file()
  ),
  height = 5,
  width = 3
)
```

## Per-gene and combined statistics

```{r}
total_genomes <- 242

per_gene_stats_dt <- rbindlist(sum_structl) %>%
  complete(contigID, software,
           fill = list(n_capsid = 0,
                       n_terminase = 0,
                       n_portal = 0)) %>%
  rename(MCP = n_capsid,
         TerL = n_terminase,
         Portal = n_portal) %>%
  pivot_longer(cols = c("MCP", "TerL", "Portal"),
               names_to = "gene",
               values_to = "count") %>%
  mutate(
    annotation_status = case_when(
      count == 0 ~ "under_annotated",
      count == 1 ~ "correct",
      count > 1 ~ "over_annotated"
    )
  ) %>%
  group_by(software, gene, annotation_status) %>%
  summarize(n_genomes = n(), .groups = "drop") %>%
  complete(software, gene, annotation_status,
           fill = list(n_genomes = 0)) %>%
  pivot_wider(names_from = annotation_status,
              values_from = n_genomes,
              values_fill = 0) %>%
  mutate(
    detection_rate = (correct + over_annotated) / total_genomes,
    accuracy_rate = correct / total_genomes,
    under_annotation_rate = under_annotated / total_genomes,
    over_annotation_rate = over_annotated / total_genomes
  )

per_gene_stats_dt
```

Combined "perfect" metric with 95% CI
```{r}
perfect_stats_dt <- rbindlist(sum_structl) %>%
  complete(contigID, software,
           fill = list(n_capsid = 0,
                       n_terminase = 0,
                       n_portal = 0)) %>%
  rename(MCP = n_capsid,
         TerL = n_terminase,
         Portal = n_portal) %>%
  mutate(perfect = MCP == 1 & TerL == 1 & Portal == 1) %>%
  group_by(software) %>%
  summarize(
    n_perfect = sum(perfect),
    n_imperfect = sum(!perfect),
    .groups = "drop"
  ) %>%
  rowwise() %>%
  mutate(
    perfect_rate = n_perfect / total_genomes,
    ci_lower = binom.test(n_perfect, total_genomes)$conf.int[1],
    ci_upper = binom.test(n_perfect, total_genomes)$conf.int[2]
  ) %>%
  ungroup()

perfect_stats_dt
```

Pairwise Fisher's exact tests between tools for "perfect" annotation
```{r}
software_list <- unique(perfect_stats_dt$software)
pairwise_results <- list()

for (i in 1:(length(software_list) - 1)) {
  for (j in (i + 1):length(software_list)) {
    sw1 <- software_list[i]
    sw2 <- software_list[j]
    
    n1_perfect <- perfect_stats_dt$n_perfect[perfect_stats_dt$software == sw1]
    n1_imperfect <- perfect_stats_dt$n_imperfect[perfect_stats_dt$software == sw1]
    n2_perfect <- perfect_stats_dt$n_perfect[perfect_stats_dt$software == sw2]
    n2_imperfect <- perfect_stats_dt$n_imperfect[perfect_stats_dt$software == sw2]
    
    contingency <- matrix(c(n1_perfect, n1_imperfect,
                            n2_perfect, n2_imperfect),
                          nrow = 2, byrow = TRUE)
    
    test_result <- fisher.test(contingency)
    
    pairwise_results[[length(pairwise_results) + 1]] <- data.table(
      software_1 = sw1,
      software_2 = sw2,
      odds_ratio = test_result$estimate,
      p_value = test_result$p.value
    )
  }
}

pairwise_fisher_dt <- rbindlist(pairwise_results) %>%
  mutate(
    p_adjusted = p.adjust(p_value, method = "BH"),
    significant = p_adjusted < 0.05
  )

pairwise_fisher_dt
```

Write results to CSV
```{r}
write.csv(
  per_gene_stats_dt,
  file = sprintf(
    "%s/data/annotation/seawater_dtrs/seawater_dtr_per_gene_stats.csv",
    find_rstudio_root_file()
  ),
  row.names = FALSE
)

write.csv(
  perfect_stats_dt,
  file = sprintf(
    "%s/data/annotation/seawater_dtrs/seawater_dtr_perfect_stats.csv",
    find_rstudio_root_file()
  ),
  row.names = FALSE
)

write.csv(
  pairwise_fisher_dt,
  file = sprintf(
    "%s/data/annotation/seawater_dtrs/seawater_dtr_pairwise_fisher.csv",
    find_rstudio_root_file()
  ),
  row.names = FALSE
)

cat("CSV files written to data/annotation/seawater_dtrs/:\n")
cat("  - seawater_dtr_per_gene_stats.csv\n")
cat("  - seawater_dtr_perfect_stats.csv\n")
cat("  - seawater_dtr_pairwise_fisher.csv\n")
```


compare non-hypothetical annotation rate
```{r}
ct3_annot_rate_dt <- merge(struct_all_dt %>% distinct(contigID),
                           ct3_merge_dt,
                           by.x = "contigID", 
                           by.y = "input_name") %>%
  filter(!grepl("tRNA-", gene_name)) %>%
  group_by(contigID) %>%
  summarize(gene_count = n(),
            hypo_annotations = n_distinct(gene_name[evidence_description == "hypothetical protein"])) %>%
  mutate(software = "Cenote-Taker 3")
```

```{r}
phar_annot_rate_dt <- merge(struct_all_dt %>% distinct(contigID),
                           phar_gene_dt,
                           by.x = "contigID", 
                           by.y = "contig") %>%
  group_by(contigID) %>%
  summarize(gene_count = n(),
            hypo_annotations = n_distinct(gene[annot == "hypothetical protein"])) %>%
  mutate(software = "Pharokka")
```


```{r}
metc_annot_rate_dt <- metc_info_dt %>%
  group_by(contigID) %>%
  summarize(gene_count = n(),
            hypo_annotations = 
              n_distinct(target[grepl("hypothetical|no_annotation", 
                                         product, 
                                         ignore.case = T)])) %>%
  mutate(software = "Metacerberus")
```


```{r}
gnmd_annot_rate_dt <- merge(struct_all_dt %>% distinct(contigID),
                           gnmd_gene_dt %>%
                             mutate(contigID = 
                                      sub("_[^_]+$", "", gene)),
                           by.x = "contigID", 
                           by.y = "contigID") %>%
  group_by(contigID) %>%
  summarize(gene_count = n(),
            hypo_annotations = 
              n_distinct(gene[is.na(annotation_description)])) %>%
  mutate(software = "geNomad")
```


```{r}
phold_annot_rate_dt <- merge(struct_all_dt %>% distinct(contigID),
                           phold_gene_dt,
                           by.x = "contigID",
                           by.y = "contig_id") %>%
  group_by(contigID) %>%
  summarize(gene_count = n(),
            hypo_annotations = 
              n_distinct(cds_id[grepl("hypothetical", 
                                         product, 
                                         ignore.case = T)])) %>%
  mutate(software = "phold")
```


combine and plot
```{r}
sum_hypol <- list(ct3_annot_rate_dt, gnmd_annot_rate_dt, 
                    metc_annot_rate_dt, phar_annot_rate_dt,
                  phold_annot_rate_dt)

hypo_all_dt <- rbindlist(sum_hypol) %>%
  mutate(hypo_prop = hypo_annotations/gene_count,
         software = fct_relevel(software,  "geNomad",
                                           "Metacerberus", 
                                           "Pharokka", 
                                           "Cenote-Taker 3",
                                           "phold")) %>%
  complete(contigID, software,
           fill = list(hypo_prop = 1))
```

```{r}
hypo_rate_p <- hypo_all_dt %>%
  ggplot(aes(x = software, y = 1 - hypo_prop,
             fill = software)) +
  geom_boxplot() + 
  geom_point(color = "grey30") +
  scale_fill_manual(values = 
                      c("#0A9F9D", "#CEB175", 
                        "#E54E21", "#6C8645", #"#C18748", 
                        "#591f6e")) + 
  stat_compare_means(
    method = "wilcox.test",
    label = "p.signif",
    comparisons = list(
      c("Cenote-Taker 3", "geNomad"),
      c("Cenote-Taker 3", "Metacerberus"),
      c("Cenote-Taker 3", "Pharokka"),
      c("Cenote-Taker 3", "phold")
    )
  ) +
  theme_bw() +
  scale_y_continuous(labels = scales::percent) +
  labs(y = "Percentage of genes with\nnon-hypothetical annotation",
       x = "",
       #title = "Seawater Virome Complete (Circular) Head-Tail Viruses, n=242"
       ) +
  theme(
    legend.position = "none",
    axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)
    )

hypo_rate_p

ggsave(
  hypo_rate_p,
  file = sprintf("%s/charts/seawater_nonhypo_box3.pdf",
                 find_rstudio_root_file()),
  height = 5,
  width = 3
)

ggsave(
  hypo_rate_p,
  file = sprintf("%s/charts/seawater_nonhypo_box3.png",
                 find_rstudio_root_file()),
  height = 5,
  width = 3
)


```

```{r}
right_side <- cowplot::plot_grid(
  essential_p, hypo_rate_p,
  ncol = 1
)
#right_side
all_p <- cowplot::plot_grid(
  struct_p, right_side,
  ncol = 2,
  rel_widths = c(2, 1)
)
all_p

ggsave(
  all_p,
  file = sprintf(
    "%s/charts/seawater_annot_combined_plots1.pdf",
    find_rstudio_root_file()
  ),
  width = 9,
  height = 8
)
```


