---
title: "Comparing Cenote-Taker 3 and Genomad Discovery on Long Circular Contigs"
output: html_notebook
---

Cenote-Taker 3 and Genomad both can discover viruses in metagenomic data.
As the field generates more long read-based metagenomic assemblies, more complete virus genomes will be assembled.
Complete sequences are highly valuable for virus cataloging (compared to incomplete genome fragments)

```{r}
# load libraries
library(data.table)
library(stringr)
library(dplyr)
library(tidyr)
library(ggplot2)
library(forcats)
library(rprojroot)
library(ggupset)
```

## Functions

```{r}
#' Tag overlapping windows (â‰¥25 % overlap) with a unique ID
#' @param df         data.frame / data.table containing window rows
#' @param contig_col column name holding contig values (default "contig")
#' @param start_col  column name holding window start (default "start")
#' @param stop_col   column name holding window stop  (default "stop")
#' @return           original data with new column `ovl_id`
assign_overlap_id <- function(df,
                              contig_col = "contig",
                              start_col  = "start",
                              stop_col   = "stop") {
  stopifnot(contig_col %in% names(df),
            start_col  %in% names(df),
            stop_col   %in% names(df))
  
  # convert to data.table for efficient in-place ops
  dt <- as.data.table(df)
  
  # standardise working names
  setnames(dt,
           c(contig_col, start_col, stop_col),
           c("contig__", "start__", "stop__"))
  
  setorder(dt, contig__, start__, stop__)
  
  # helper to tag one contig cluster vectorised over rows
  tag_one <- function(start_vec, stop_vec) {
    n <- length(start_vec)
    grp <- integer(n)
    cur_tag <- 1L
    cur_s <- start_vec[1]
    cur_e <- stop_vec[1]
    cur_len <- cur_e - cur_s
    grp[1] <- cur_tag
    if (n > 1L) {
      for (i in 2:n) {
        len_i <- stop_vec[i] - start_vec[i]
        ovl   <- min(cur_e, stop_vec[i]) - max(cur_s, start_vec[i])
        if (ovl > 0 &&
            ovl >= 0.25 * pmin(len_i, cur_len)) {
          grp[i] <- cur_tag
          if (stop_vec[i] > cur_e) {
            cur_e  <- stop_vec[i]
            cur_len <- cur_e - cur_s
          }
        } else {
          cur_tag <- cur_tag + 1L
          grp[i]  <- cur_tag
          cur_s <- start_vec[i]
          cur_e <- stop_vec[i]
          cur_len <- cur_e - cur_s
        }
      }
    }
    grp
  }
  
  dt[, ovl_grp := tag_one(start__, stop__), by = contig__]
  dt[, ovl_id  := paste0(contig__, "_", ovl_grp)]
  
  # restore original column names
  setnames(dt,
           c("contig__", "start__", "stop__"),
           c(contig_col, start_col, stop_col))
  
  invisible(dt[])
}

```

## DRR290133

load tables
```{r}
DRR290133_ct_sum_dt <- fread(
  sprintf(
    "%s/data/discovery/DRR290133/ct3/ct3_DRR290133_1b_virus_summary.tsv",
    find_rstudio_root_file()
  ),
  header = T
)

DRR290133_ct3_prune_dt <- fread(
  sprintf(
    "%s/data/discovery/DRR290133/ct3/ct3_DRR290133_1b_prune_summary.tsv",
    find_rstudio_root_file()
  ),
  header = T
) %>%
  mutate(chunk_ID = paste(contig, chunk_name, sep = "@")) %>%
  select(c(chunk_ID, contig_length, chunk_start, chunk_stop))

DRR290133_ct_merge_dt <- merge(
  DRR290133_ct_sum_dt,
  DRR290133_ct3_prune_dt,
  by.x = "contig",
  by.y = "chunk_ID",
  all.x = T
) %>%
  mutate(
    chunk_start = case_when(
      is.na(chunk_start) ~ 0,
      TRUE ~ chunk_start
    ),
    chunk_stop = case_when(
      is.na(chunk_stop) ~ virus_seq_length,
      TRUE ~ chunk_stop
    ),
    input_name = gsub(" \\(.*", "", input_name)
  )

```

```{r}
DRR290133_genomad_dt <- fread(
  sprintf(
    "%s/data/discovery/DRR290133/genomad/DRR290133_myloasm_1a.circ_with_dtr_virus_summary.tsv",
    find_rstudio_root_file()
  ),
  header = T
) %>% 
  separate(seq_name, 
           sep = "\\|",
           c("input_name", "provirus")) %>%
  separate(coordinates, 
           sep = "-",
           c("chunk_start", "chunk_stop")) %>%
  mutate(chunk_start = as.double(chunk_start),
         chunk_stop = as.double(chunk_stop)) %>%
  mutate(
    chunk_start = case_when(
      is.na(chunk_start) ~ 0,
      TRUE ~ chunk_start
    ),
    chunk_stop = case_when(
      is.na(chunk_stop) ~ length,
      TRUE ~ chunk_stop
    )
  )
```

```{r}
DRR290133_full_dt <- merge(
  DRR290133_ct_merge_dt,
  DRR290133_genomad_dt,
  by = c("input_name", "chunk_start", "chunk_stop"),
  all = T
)
```

```{r}
DRR290133_view_dt <- DRR290133_full_dt %>%
  select(c(input_name, chunk_start, chunk_stop,
           organism, virion_hallmark_count,
           n_hallmarks, marker_enrichment))
```

```{r}
DRR290133_view2_dt <- assign_overlap_id(
  DRR290133_view_dt,
  contig_col = "input_name",
  start_col = "chunk_start",
  stop_col = "chunk_stop"
) %>%
  group_by(ovl_id) %>%
  summarize(
    approx_len = max(chunk_stop) - min(chunk_start),
    min_start = min(chunk_start),
    max_stop = max(chunk_stop),
    organism = first(organism, na_rm = T),
    virion_hallmark_count = first(virion_hallmark_count, na_rm = T),
    n_hallmarks = first(n_hallmarks, na_rm = T),
    marker_enrichment = first(marker_enrichment, na_rm = T),
  )
```

Cenote-Taker only
```{r}
DRR290133_view2_dt %>%
  filter(is.na(n_hallmarks))
```

```{r}
DRR290133_ct3_only_dt <- DRR290133_view2_dt %>%
  filter(is.na(n_hallmarks)) %>%
  mutate(
    input_name = gsub("_[0-9]$", "", ovl_id)
  ) %>%
  rename(
    chunk_start = min_start,
    chunk_stop = max_stop
  ) %>%
  select(c(input_name, chunk_start, chunk_stop))

DRR290133_ct3_contigs <- merge(
  DRR290133_ct3_only_dt,
  DRR290133_ct_merge_dt,
  by = c("input_name", "chunk_start", "chunk_stop")
) %>%
  select(c(input_name, contig))

write.table(
  DRR290133_ct3_contigs$contig,
  file = sprintf(
    "%s/data/discovery/DRR290133/ct3/ct3_only_contigs.txt",
    find_rstudio_root_file()
  ),
  col.names = F,
  row.names = F,
  quote = F
)

```

Genomad contigs only
```{r}
DRR290133_view2_dt %>%
  filter(is.na(virion_hallmark_count))
```

```{r}
DRR290133_genomad_only_dt <- DRR290133_view2_dt %>%
  filter(is.na(virion_hallmark_count)) %>%
  mutate(
    input_name = gsub("_[0-9]$", "", ovl_id)
  ) %>%
  rename(
    chunk_start = min_start,
    chunk_stop = max_stop
  ) %>%
  select(c(input_name, chunk_start, chunk_stop))

DRR290133_genomad_contigs <- merge(
  DRR290133_genomad_only_dt,
  DRR290133_genomad_dt,
  by = c("input_name", "chunk_start", "chunk_stop")
) %>%
  select(c(input_name, provirus)) %>%
  mutate(
    contig = paste(input_name, provirus, sep = "|"),
    contig = gsub("\\|NA", "", contig)
    )

write.table(
  DRR290133_genomad_contigs$contig,
  file = sprintf(
    "%s/data/discovery/DRR290133/genomad/genomad_only_contigs.txt",
    find_rstudio_root_file()
  ),
  col.names = F,
  row.names = F,
  quote = F
)

```



```{r}
DRR290133_view2_dt %>%
  filter(!is.na(virion_hallmark_count),
         !is.na(n_hallmarks)
         )
```

upset plot
```{r}

DRR290133_upset_ct_dt <- assign_overlap_id(
  DRR290133_view_dt,
  contig_col = "input_name",
  start_col = "chunk_start",
  stop_col = "chunk_stop"
) %>%
  group_by(ovl_id) %>%
  summarize(
    hit = sum(!is.na(organism)),
  ) %>%
  ungroup() %>%
  filter(hit >= 1) %>%
  mutate(software = "Cenote-Taker 3")

DRR290133_upset_gnmd_dt <- assign_overlap_id(
  DRR290133_view_dt,
  contig_col = "input_name",
  start_col = "chunk_start",
  stop_col = "chunk_stop"
) %>%
  group_by(ovl_id) %>%
  summarize(
    hit = sum(!is.na(marker_enrichment)),
  ) %>%
  ungroup() %>%
  filter(hit >= 1) %>%
  mutate(software = "geNomad")


DRR290133_upset_dt <- rbind(
  DRR290133_upset_ct_dt,
  DRR290133_upset_gnmd_dt
) %>%
  group_by(ovl_id) %>%
  summarize(softwares = list(software))

DRR290133_up_p <- DRR290133_upset_dt %>%
  ggplot(aes(x= softwares)) +
  geom_bar(color = "black", fill = "lightgrey") +
  scale_x_upset() +
  geom_text(stat='count', aes(label=after_stat(count)), vjust=+1.1) +
  labs(
    x = "Software",
    y = "Predicted Viruses"
  ) +
  theme_bw()

DRR290133_up_p

ggsave(
  DRR290133_up_p,
  file = sprintf(
    "%s/charts/DRR290133_disc_upset1.pdf",
    find_rstudio_root_file()
  ),
  width = 3,
  height = 4
)

```

DRR290133 genomad only stats
```{r}
DRR_gen_hall_p <- merge(
  DRR290133_genomad_dt,
  DRR290133_genomad_only_dt %>% mutate(g_only = "geNomad-only"),
  all.x = T
) %>%
  mutate(g_only = case_when(g_only == "geNomad-only" ~ "geNomad-only", TRUE ~ "CT3 and geNomad")) %>%
  ggplot(aes(x=n_hallmarks)) +
  geom_histogram() +
  facet_wrap(~g_only, ncol = 1) +
  theme_bw() +
  labs(
    x = "geNomad hallmark gene count",
    y = "genome count",
    title = "geNomad hallmark genes/virus MAG"
  )

DRR_gen_hall_p

ggsave(
  DRR_gen_hall_p,
  file = sprintf(
    "%s/charts/DRR290133_disc_genomad_halls1.pdf",
    find_rstudio_root_file()
  ),
  width = 4,
  height = 4
)

ggsave(
  DRR_gen_hall_p,
  file = sprintf(
    "%s/charts/DRR290133_disc_genomad_halls1.png",
    find_rstudio_root_file()
  ),
  width = 4,
  height = 4
)
```


DRR290133 CT3 only stats
```{r}
DRR_ct3_hall_p <- merge(
  DRR290133_ct3_only_dt %>% mutate(c_only = "CT3-only"),
  DRR290133_ct_merge_dt,
  all.y = T
) %>%
  mutate(
    c_only = case_when(c_only == "CT3-only" ~ "CT3-only", TRUE ~ "CT3 and geNomad"),
    n_hallmarks = virion_hallmark_count + rep_hallmark_count
    ) %>%
  ggplot(aes(x=n_hallmarks)) +
  geom_histogram() +
  facet_wrap(~c_only, ncol = 1) +
  theme_bw() +
  labs(
    x = "CT3 hallmark gene count",
    y = "CT3 count",
    title = "CT3 hallmark genes/virus MAG"
  )

DRR_ct3_hall_p

ggsave(
  DRR_ct3_hall_p,
  file = sprintf(
    "%s/charts/DRR290133_disc_ct3_halls1.pdf",
    find_rstudio_root_file()
  ),
  width = 4,
  height = 4
)

ggsave(
  DRR_ct3_hall_p,
  file = sprintf(
    "%s/charts/DRR290133_disc_ct3_halls1.png",
    find_rstudio_root_file()
  ),
  width = 4,
  height = 4
)

```



## ERR10905741

load tables
```{r}
ERR10905741_ct_sum_dt <- fread(
  sprintf(
    "%s/data/discovery/ERR10905741/ct3/ct3_ERR10905741_1b_virus_summary.tsv",
    find_rstudio_root_file()
  ),
  header = T
)

ERR10905741_ct3_prune_dt <- fread(
  sprintf(
    "%s/data/discovery/ERR10905741/ct3/ct3_ERR10905741_1b_prune_summary.tsv",
    find_rstudio_root_file()
  ),
  header = T
) %>%
  mutate(chunk_ID = paste(contig, chunk_name, sep = "@")) %>%
  select(c(chunk_ID, contig_length, chunk_start, chunk_stop))

ERR10905741_ct_merge_dt <- merge(
  ERR10905741_ct_sum_dt,
  ERR10905741_ct3_prune_dt,
  by.x = "contig",
  by.y = "chunk_ID",
  all.x = T
) %>%
  mutate(
    chunk_start = case_when(
      is.na(chunk_start) ~ 0,
      TRUE ~ chunk_start
    ),
    chunk_stop = case_when(
      is.na(chunk_stop) ~ virus_seq_length,
      TRUE ~ chunk_stop
    ),
    input_name = gsub(" \\(.*", "", input_name)
  )

```

```{r}
ERR10905741_genomad_dt <- fread(
  sprintf(
    "%s/data/discovery/ERR10905741/genomad/ERR10905741_myloasm_1a.circ_with_dtr_virus_summary.tsv",
    find_rstudio_root_file()
  ),
  header = T
) %>% 
  separate(seq_name, 
           sep = "\\|",
           c("input_name", "provirus")) %>%
  separate(coordinates, 
           sep = "-",
           c("chunk_start", "chunk_stop")) %>%
  mutate(chunk_start = as.double(chunk_start),
         chunk_stop = as.double(chunk_stop)) %>%
  mutate(
    chunk_start = case_when(
      is.na(chunk_start) ~ 0,
      TRUE ~ chunk_start
    ),
    chunk_stop = case_when(
      is.na(chunk_stop) ~ length,
      TRUE ~ chunk_stop
    )
  )
```

```{r}
ERR10905741_full_dt <- merge(
  ERR10905741_ct_merge_dt,
  ERR10905741_genomad_dt,
  by = c("input_name", "chunk_start", "chunk_stop"),
  all = T
)
```

```{r}
ERR10905741_view_dt <- ERR10905741_full_dt %>%
  select(c(input_name, chunk_start, chunk_stop,
           organism, virion_hallmark_count,
           n_hallmarks, marker_enrichment))

 
```

```{r}
ERR10905741_view2_dt <- assign_overlap_id(
  ERR10905741_view_dt,
  contig_col = "input_name",
  start_col = "chunk_start",
  stop_col = "chunk_stop"
) %>%
  group_by(ovl_id) %>%
  summarize(
    approx_len = max(chunk_stop) - min(chunk_start),
    min_start = min(chunk_start),
    max_stop = max(chunk_stop),
    organism = first(organism, na_rm = T),
    virion_hallmark_count = first(virion_hallmark_count, na_rm = T),
    n_hallmarks = first(n_hallmarks, na_rm = T),
    marker_enrichment = first(marker_enrichment, na_rm = T),
  )
```

Cenote-Taker only
```{r}
ERR10905741_view2_dt %>%
  filter(is.na(n_hallmarks))
```

```{r}
ERR10905741_ct3_only_dt <- ERR10905741_view2_dt %>%
  filter(is.na(n_hallmarks)) %>%
  mutate(
    input_name = gsub("_[0-9]$", "", ovl_id)
  ) %>%
  rename(
    chunk_start = min_start,
    chunk_stop = max_stop
  ) %>%
  select(c(input_name, chunk_start, chunk_stop))

ERR10905741_ct3_contigs <- merge(
  ERR10905741_ct3_only_dt,
  ERR10905741_ct_merge_dt,
  by = c("input_name", "chunk_start", "chunk_stop")
) %>%
  select(c(input_name, contig))

write.table(
  ERR10905741_ct3_contigs$contig,
  file = sprintf(
    "%s/data/discovery/ERR10905741/ct3/ct3_only_contigs.txt",
    find_rstudio_root_file()
  ),
  col.names = F,
  row.names = F,
  quote = F
)

```

Genomad contigs only
```{r}
ERR10905741_view2_dt %>%
  filter(is.na(virion_hallmark_count))
```

```{r}
ERR10905741_genomad_only_dt <- ERR10905741_view2_dt %>%
  filter(is.na(virion_hallmark_count)) %>%
  mutate(
    input_name = gsub("_[0-9]$", "", ovl_id)
  ) %>%
  rename(
    chunk_start = min_start,
    chunk_stop = max_stop
  ) %>%
  select(c(input_name, chunk_start, chunk_stop))

ERR10905741_genomad_contigs <- merge(
  ERR10905741_genomad_only_dt,
  ERR10905741_genomad_dt,
  by = c("input_name", "chunk_start", "chunk_stop")
) %>%
  select(c(input_name, provirus)) %>%
  mutate(
    contig = paste(input_name, provirus, sep = "|"),
    contig = gsub("\\|NA", "", contig)
    )

write.table(
  ERR10905741_genomad_contigs$contig,
  file = sprintf(
    "%s/data/discovery/ERR10905741/genomad/genomad_only_contigs.txt",
    find_rstudio_root_file()
  ),
  col.names = F,
  row.names = F,
  quote = F
)

```

```{r}
ERR10905741_view2_dt %>%
  filter(!is.na(virion_hallmark_count),
         !is.na(n_hallmarks)
         )
```

upset plot
```{r}

ERR10905741_upset_ct_dt <- assign_overlap_id(
  ERR10905741_view_dt,
  contig_col = "input_name",
  start_col = "chunk_start",
  stop_col = "chunk_stop"
) %>%
  group_by(ovl_id) %>%
  summarize(
    hit = sum(!is.na(organism)),
  ) %>%
  ungroup() %>%
  filter(hit >= 1) %>%
  mutate(software = "Cenote-Taker 3")

ERR10905741_upset_gnmd_dt <- assign_overlap_id(
  ERR10905741_view_dt,
  contig_col = "input_name",
  start_col = "chunk_start",
  stop_col = "chunk_stop"
) %>%
  group_by(ovl_id) %>%
  summarize(
    hit = sum(!is.na(marker_enrichment)),
  ) %>%
  ungroup() %>%
  filter(hit >= 1) %>%
  mutate(software = "geNomad")


ERR10905741_upset_dt <- rbind(
  ERR10905741_upset_ct_dt,
  ERR10905741_upset_gnmd_dt
) %>%
  group_by(ovl_id) %>%
  summarize(softwares = list(software))

ERR10905741_up_p <- ERR10905741_upset_dt %>%
  ggplot(aes(x= softwares)) +
  geom_bar(color = "black", fill = "lightgrey") +
  scale_x_upset() +
  geom_text(stat='count', aes(label=after_stat(count)), vjust=+1.1) +
  labs(
    x = "Software",
    y = "Predicted Viruses"
  ) +
  theme_bw()

ERR10905741_up_p

ggsave(
  ERR10905741_up_p,
  file = sprintf(
    "%s/charts/ERR10905741_disc_upset1.pdf",
    find_rstudio_root_file()
  ),
  width = 3,
  height = 4
)

```
ERR1095741 genomad only stats
```{r}
ERR_gen_hall_p <- merge(
  ERR10905741_genomad_dt,
  ERR10905741_genomad_only_dt %>% mutate(g_only = "geNomad-only"),
  all.x = T
) %>%
  mutate(g_only = case_when(g_only == "geNomad-only" ~ "geNomad-only", TRUE ~ "CT3 and geNomad")) %>%
  ggplot(aes(x=n_hallmarks)) +
  geom_histogram() +
  facet_wrap(~g_only, ncol = 1) +
  theme_bw() +
  labs(
    x = "geNomad hallmark gene count",
    y = "genome count",
    title = "geNomad hallmark genes/virus MAG"
  )

ERR_gen_hall_p

ggsave(
  ERR_gen_hall_p,
  file = sprintf(
    "%s/charts/ERR10905741_disc_genomad_halls1.pdf",
    find_rstudio_root_file()
  ),
  width = 4,
  height = 4
)

ggsave(
  ERR_gen_hall_p,
  file = sprintf(
    "%s/charts/ERR10905741_disc_genomad_halls1.png",
    find_rstudio_root_file()
  ),
  width = 4,
  height = 4
)
```


ERR1095741 CT3 only stats
```{r}
ERR_ct3_hall_p <- merge(
  ERR10905741_ct3_only_dt %>% mutate(c_only = "CT3-only"),
  ERR10905741_ct_merge_dt,
  all.y = T
) %>%
  mutate(
    c_only = case_when(c_only == "CT3-only" ~ "CT3-only", TRUE ~ "CT3 and geNomad"),
    n_hallmarks = virion_hallmark_count + rep_hallmark_count
    ) %>%
  ggplot(aes(x=n_hallmarks)) +
  geom_histogram() +
  facet_wrap(~c_only, ncol = 1) +
  theme_bw() +
  labs(
    x = "CT3 hallmark gene count",
    y = "CT3 count",
    title = "CT3 hallmark genes/virus MAG"
  )

ERR_ct3_hall_p

ggsave(
  ERR_ct3_hall_p,
  file = sprintf(
    "%s/charts/ERR10905741_disc_ct3_halls1.pdf",
    find_rstudio_root_file()
  ),
  width = 4,
  height = 4
)

ggsave(
  ERR_ct3_hall_p,
  file = sprintf(
    "%s/charts/ERR10905741_disc_ct3_halls1.png",
    find_rstudio_root_file()
  ),
  width = 4,
  height = 4
)

```

```{r}
merge(
  ERR10905741_genomad_dt,
  ERR10905741_genomad_only_dt
) %>%
  filter(n_hallmarks == 0)
```

